filemanager/s3manager_v2.go

```diff
@@ -0,0 +1,366 @@
+) (*S3ManagerV2, error) {
+	var s3Config S3Config
+	if err := mapstructure.Decode(config, &s3Config); err != nil {
+		return nil, fmt.Errorf("failed to decode S3 config: %w", err)
+	}
+
+	sessionConfig, err := awsutil.NewSimpleSessionConfig(config, ServiceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AWS session config: %w", err)
+	}
+
+	s3Config.RegionHint = appConfig.GetString("AWS_S3_REGION_HINT", "us-east-1")
+
+	return &S3ManagerV2{
+		baseManager: &baseManager{
+			logger:         log,
+			defaultTimeout: defaultTimeout,
+		},
+		config:        &s3Config,
+		sessionConfig: sessionConfig,
+	}, nil
+}
+
+// ListFilesWithPrefix returns a session for listing files with the given prefix.
+func (m *S3ManagerV2) ListFilesWithPrefix(ctx context.Context, startAfter, prefix string, maxItems int64) ListSession {
+	return &s3ListSessionV2{
+		baseListSession: &baseListSession{
+			ctx:        ctx,
+			startAfter: startAfter,
+			prefix:     prefix,
+			maxItems:   maxItems,
+		},
+		manager:     m,
+		isTruncated: true,
+	}
+}
+
+// Download downloads a file from S3 to the provided io.WriterAt.
+func (m *S3ManagerV2) Download(ctx context.Context, output io.WriterAt, key string) error {
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return fmt.Errorf("s3 client: %w", err)
+	}
+
+	downloader := s3manager.NewDownloader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	_, err = downloader.Download(ctx, output,
+		&s3.GetObjectInput{
+			Bucket: aws.String(m.config.Bucket),
+			Key:    aws.String(key),
+		})
+	if err == nil {
+		return nil
+	}
+	var nsk *types.NoSuchKey
+	if errors.As(err, &nsk) {
+		return ErrKeyNotFound
+	}
+	return fmt.Errorf("failed to download from S3: %w", err)
+}
+
+// Upload uploads a file to S3 and returns the uploaded file info.
+func (m *S3ManagerV2) Upload(ctx context.Context, file *os.File, prefixes ...string) (UploadedFile, error) {
+	fileName := path.Join(m.config.Prefix, path.Join(prefixes...), path.Base(file.Name()))
+
+	uploadInput := &s3.PutObjectInput{
+		ACL:    types.ObjectCannedACLBucketOwnerFullControl,
+		Bucket: aws.String(m.config.Bucket),
+		Key:    aws.String(fileName),
+		Body:   file,
+	}
+
+	if m.config.EnableSSE {
+		uploadInput.ServerSideEncryption = types.ServerSideEncryptionAes256
+	}
+
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return UploadedFile{}, fmt.Errorf("s3 client: %w", err)
+	}
+	uploader := s3manager.NewUploader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	output, err := uploader.Upload(ctx, uploadInput)
+	if err == nil {
+		return UploadedFile{Location: output.Location, ObjectName: fileName}, nil
+	}
+	var regionError *aws.MissingRegionError
+	if errors.As(err, &regionError) {
+		err = fmt.Errorf(`missing region for bucket %q: %w`, m.config.Bucket, regionError)
+	}
+
+	return UploadedFile{}, err
```

Comment:
why not as it was before?
```suggestion
	objName := path.Join(m.config.Prefix, path.Join(prefixes...), path.Base(file.Name()))
	return m.UploadReader(ctx, objName, file)
```